### PR TITLE
Abdul/token accounts

### PR DESCRIPTION
**Pull request by imabdulbasit**
- rebased token_accounts
- added token_transfers
-

### CODE REVIEW

File: crates/indexer/src/accountsdb/programs/token.rs [hunk 1]
```diff
@@ -1,6 +1,11 @@
...
+use spl_token::state::Account as TokenAccount;
+
+use super::accounts::token;
 use crate::{prelude::*, Client};
 
-pub(crate) async fn process(_client: &Client, _key: Pubkey, _data: Vec<u8>) -> Result<()> {
-    // TODO
+pub(crate) async fn process(client: &Client, key: Pubkey, data: Vec<u8>) -> Result<()> {
+    let token_account = TokenAccount::unpack_unchecked(&data).context("failed!")?;
+    token::process(client, key, token_account).await;
```
Comment (ray-kast) [inline crates/indexer/src/accountsdb/programs/token.rs:9]:
```suggestion
    token::process(client, key, token_account).await
```

File: crates/indexer/src/accountsdb/accounts/token.rs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+                .values(&values)
+                .on_conflict(token_accounts::address)
+                .do_update()
+                .set(&values)
+                .execute(db)
+        })
+        .await
+        .context("failed to insert token account")?;
+    process_token_transfer(client, key, token_account).await;
+    Ok(())
```
Comment (ray-kast) [inline crates/indexer/src/accountsdb/accounts/token.rs:101]:
```suggestion
    process_token_transfer(client, key, token_account).await
```

File: crates/indexer/src/accountsdb/accounts/token.rs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+pub async fn process(client: &Client, key: Pubkey, token_account: TokenAccount) -> Result<()> {
+    let mint = token_account.mint.to_string();
+    let res = client
+        .db(move |db| {
+            metadatas::table
+                .filter(metadatas::mint_address.eq(&mint))
+                .execute(db)
+        })
+        .await
+        .context("Failed to check mint address for existing mint")?;
```
Comment (ray-kast) [inline crates/indexer/src/accountsdb/accounts/token.rs:69]: This should use a `SELECT EXISTS` query

File: crates/indexer/src/accountsdb/accounts/token.rs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    db::{
+        insert_into,
+        models::{TokenAccount as TokenAccountModel, TokenTransfer},
+        select,
+        tables::{metadatas, token_accounts, token_transfers},
+        update,
+    },
+    prelude::*,
+};
+use solana_program::program_pack::Pack;
```
Comment (ray-kast) [inline crates/indexer/src/accountsdb/accounts/token.rs:12]:
```suggestion
```

File: crates/indexer/src/accountsdb/accounts/token.rs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+) -> Result<()> {
+    let accts = client
+        .db(move |db| {
+            token_accounts::table
+                .filter(token_accounts::address.eq(&key.to_string()))
+                .load::<TokenAccountModel>(db)
+        })
+        .await
+        .context("Failed to check token_accounts!")?;
+    let acct = accts.get(0).context("index out of bound!")?;
```
Comment (ray-kast) [inline crates/indexer/src/accountsdb/accounts/token.rs:30]:
```suggestion
    let acct = accts.get(0).context("Could not find existing token account")?;
```

**Review by ray-kast — CHANGES_REQUESTED** [review]

File: crates/core/migrations/2022-02-12-065030_add_not_null_contraint_to_token_account_amount/down.sql [hunk 6]
```diff
@@ -0,0 +1 @@
+ALTER TABLE token_accounts ALTER COLUMN amount DROP not null;
```
Comment (ray-kast) [inline crates/core/migrations/2022-02-12-065030_add_not_null_contraint_to_token_account_amount/down.sql:1]:
```suggestion
alter table token_accounts
alter column amount drop not null;
```

File: crates/core/migrations/2022-02-12-065030_add_not_null_contraint_to_token_account_amount/up.sql [hunk 7]
```diff
@@ -0,0 +1 @@
+ALTER TABLE token_accounts ALTER COLUMN amount SET not null;
```
Comment (ray-kast) [inline crates/core/migrations/2022-02-12-065030_add_not_null_contraint_to_token_account_amount/up.sql:1]:
```suggestion
alter table token_accounts
alter column amount set not null;
```

**Review by ray-kast — CHANGES_REQUESTED** [review]
Looks just about good to go.

File: crates/core/migrations/2022-02-09-171722_create_token_transfers_table/up.sql [hunk 8]
```diff
@@ -0,0 +1,9 @@
+create table token_transfers (
+  owner_from      varchar(48),     
+  to_owner        varchar(48),  
+  mint_address    varchar(48),     
+  transfered_at   timestamp not null,
+
+  primary key (owner_from,to_owner,mint_address,transfered_at)
```
Comment (ray-kast) [inline crates/core/migrations/2022-02-09-171722_create_token_transfers_table/up.sql:7]:
```suggestion
  owner_from      varchar(48),     
  owner_to        varchar(48),  
  mint_address    varchar(48),     
  transfered_at   timestamp not null,

  primary key (owner_from, owner_to, mint_address, transfered_at)
```

File: crates/core/migrations/2022-02-12-065030_add_not_null_contraint_to_token_account_amount/down.sql [hunk 9]
```diff
@@ -0,0 +1 @@
+alter table token_accounts alter column amount drop not null;
```
Comment (ray-kast) [inline crates/core/migrations/2022-02-12-065030_add_not_null_contraint_to_token_account_amount/down.sql:1]:
```suggestion
alter table token_accounts
alter column amount drop not null;
```

File: crates/core/migrations/2022-02-12-065030_add_not_null_contraint_to_token_account_amount/up.sql [hunk 10]
```diff
@@ -0,0 +1 @@
+alter table token_accounts alter column amount set not null;
```
Comment (ray-kast) [inline crates/core/migrations/2022-02-12-065030_add_not_null_contraint_to_token_account_amount/up.sql:1]:
```suggestion
alter table token_accounts
alter column amount set not null;
```

File: crates/core/src/db/models.rs [hunk 11]
```diff
@@ -442,3 +442,17 @@ pub struct AuctionHouse<'a> {
...
     pub can_change_sale_price: bool,
 }
+
+/// A row in the `token_transfers` table
+#[derive(Debug, Clone, Queryable, Insertable, AsChangeset)]
+pub struct TokenTransfer<'a> {
+    /// Address of the wallet from which NFT was transfered
+    pub owner_from: Cow<'a, str>,
+    /// Address of the wallet to which NFT was transfered
+    pub to_owner: Cow<'a, str>,
```
Comment (ray-kast) [inline crates/core/src/db/models.rs:452]:
```suggestion
    pub owner_to: Cow<'a, str>,
```

File: crates/core/src/db/models.rs [hunk 12]
```diff
@@ -442,3 +442,17 @@ pub struct AuctionHouse<'a> {
     /// Helpful if buyer lists an NFT with price of 0
     pub can_change_sale_price: bool,
 }
+
+/// A row in the `token_transfers` table
+#[derive(Debug, Clone, Queryable, Insertable, AsChangeset)]
+pub struct TokenTransfer<'a> {
+    /// Address of the wallet from which NFT was transfered
```
Comment (ray-kast) [inline crates/core/src/db/models.rs:449]:
```suggestion
    /// Address of the wallet from which NFT was transferred
```

File: crates/core/src/db/models.rs [hunk 13]
```diff
@@ -442,3 +442,17 @@ pub struct AuctionHouse<'a> {
...
+    pub owner_from: Cow<'a, str>,
+    /// Address of the wallet to which NFT was transfered
```
Comment (ray-kast) [inline crates/core/src/db/models.rs:451]:
```suggestion
    /// Address of the wallet to which NFT was transferred
```

File: crates/core/src/db/models.rs [hunk 14]
```diff
@@ -442,3 +442,17 @@ pub struct AuctionHouse<'a> {
...
+    pub to_owner: Cow<'a, str>,
+    /// Mint address of the token
+    pub mint_address: Cow<'a, str>,
+    /// Time at which transfered occured
+    /// this is an approximate time so the block time of the transaction signature can be different
```
Comment (ray-kast) [inline crates/core/src/db/models.rs:456]:
```suggestion
    /// Time at which transfer occurred
    ///
    /// This is an approximate time so the block time of the transaction
    /// signature can be different.
```

**Review by kespinola — APPROVED** [review]

File: crates/core/src/db/models.rs [hunk 15]
```diff
@@ -442,3 +442,17 @@ pub struct AuctionHouse<'a> {
...
+    pub transfered_at: NaiveDateTime,
```
Comment (ray-kast) [inline crates/core/src/db/models.rs:457]:
```suggestion
    pub transferred_at: NaiveDateTime,
```

File: crates/core/migrations/2022-02-09-171722_create_token_transfers_table/up.sql [hunk 16]
```diff
@@ -0,0 +1,9 @@
+create table token_transfers (
+  owner_from      varchar(48),     
+  to_owner        varchar(48),  
+  mint_address    varchar(48),     
+  transfered_at   timestamp not null,
```
Comment (ray-kast) [inline crates/core/migrations/2022-02-09-171722_create_token_transfers_table/up.sql:5]:
```suggestion
  transferred_at   timestamp not null,
```

File: crates/indexer/src/accountsdb/accounts/token.rs [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+        })
+        .await
+        .context("Failed to check token_accounts!")?;
+    let acct = accts
+        .get(0)
+        .context("Could not find existing token account!")?;
+    if token_account.owner.to_string() != acct.owner_address {
+        let row = TokenTransfer {
+            owner_from: Owned(acct.owner_address.to_string()),
+            to_owner: Owned(token_account.owner.to_string()),
```
Comment (ray-kast) [inline crates/indexer/src/accountsdb/accounts/token.rs:37]:
```suggestion
            owner_to: Owned(token_account.owner.to_string()),
```

File: crates/indexer/src/accountsdb/accounts/token.rs [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+            mint_address: Owned(token_account.mint.to_string()),
+            transfered_at: Local::now().naive_utc(),
```
Comment (ray-kast) [inline crates/indexer/src/accountsdb/accounts/token.rs:39]:
```suggestion
            transferred_at: Local::now().naive_utc(),
```

**Review by ray-kast — CHANGES_REQUESTED** [review]
Just some formatting and nitpicks

File: crates/core/migrations/2022-02-12-065030_add_not_null_contraint_to_token_account_amount/down.sql (hunk 9 above)
Comment (imabdulbasit) [reply crates/core/migrations/2022-02-12-065030_add_not_null_contraint_to_token_account_amount/down.sql:1]: well the migration won't run if I format like this :D

Comment (ray-kast) [reply crates/core/migrations/2022-02-12-065030_add_not_null_contraint_to_token_account_amount/down.sql]: @imabdulbasit that's weird, i've done it like that before in the other migrations

**Comment by ray-kast** [comment]
Re: the alter table migration, see [here](https://github.com/holaplex/solana-indexer/blob/2955ca6cb3275c1e9ae9ec59ed58ea8777a86b81/crates/core/migrations/2021-11-29-223001_remove_unnecessary_nullable/up.sql#L9-L10) and [here](https://github.com/holaplex/solana-indexer/blob/2955ca6cb3275c1e9ae9ec59ed58ea8777a86b81/crates/core/migrations/2021-11-29-223001_remove_unnecessary_nullable/down.sql#L9-L10).  Lmk if you still have trouble running the migrations.

File: crates/core/migrations/2022-02-12-065030_add_not_null_contraint_to_token_account_amount/down.sql (hunk 9 above)
Comment (imabdulbasit) [reply crates/core/migrations/2022-02-12-065030_add_not_null_contraint_to_token_account_amount/down.sql]: Ok it works now after I recreated the db

**Review by ray-kast — APPROVED** [review]

**Pull request merged**